feat: refactor zksync-hardhat documentation (#180)
matter-labs/zksync-docs: content/00.build/40.tooling/20.hardhat/00.introduction/02.installation.md

---
title: Installation
description: How to install zksync-hardhat for your project.
tags: ["zksync", "hardhart", "plugins", "installation"]
---

Before exploring ZKsync development with Hardhat, ensure your environment is properly set up. Follow these steps to install the necessary tools and
meet the prerequisites for a smooth development experience.

### Prerequisites

**System requirements:**

- Ensure your machine meets the [system requirements](%%zk_git_repo_era-compiler-solidity%%/tree/main#system-requirements) for running ZKsync and
Hardhat. For Windows users, using Windows Subsystem for Linux (WSL 2) is highly recommended for better compatibility.

**Node.js and package manager:**

> - Install [Node.js](https://nodejs.org) if it's not already installed. We recommend using a version supported by Hardhat.

Reviewer: ```suggestion
- [Node.js](https://nodejs.org) version 18 or newer. 
```

> - Choose a package manager: [Yarn](https://classic.yarnpkg.com/en/docs/install) or [npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm).

Reviewer: We can remove this line. NPM comes with node so a package manager is already available.

**Familiarity with Ethereum development:**

- Basic knowledge of deploying smart contracts on Ethereum is beneficial. Refer to
   the [Contract Deployment](/build/developer-reference/ethereum-differences/contract-deployment) documentation if you need a refresher.

**Wallet and testnet funds:**

> - You’ll need a wallet with sufficient Sepolia ETH to deploy contracts. You can obtain Sepolia ETH
>    from [network faucets](https://faucets.chain.link/sepolia) and bridge funds to ZKsync Era using [bridges](https://zksync.io/explore#bridges).

Reviewer: ```suggestion
- You’ll need a wallet with sufficient ETH to deploy contracts. 
  See the [Network Faucets](/ecosystem/network-faucets) page to find a list of resources to obtain testnet funds.
```

> - Ensure you know how to [export your private key from MetaMask](https://support.metamask.io/hc/en-us/articles/360015289632-How-to-export-an-account-s-private-key).

Reviewer: Remove this line. Not everyone uses Metamask.

Author: That was added by the developers...

Reviewer: Since we're going through making changes to the documents, we want to make improvements following our documentation styleguide. I am okay with removing this line.

::callout{icon="i-heroicons-light-bulb"}
If you're migrating an existing hardhat to `zksync-era`, please refer to the [migration guide](/build/tooling/hardhat/guides/migrating-to-zksync) for
detailed instructions.
::

### Installation steps

**Install ZKsync CLI:**

- The ZKsync CLI simplifies project setup and local testing. Install it globally with npm or yarn:

::code-group

  ```bash [yarn]
    yarn global add zksync-cli
  ```

  ```bash [npm]
    npm install -g zksync-cli
  ```

> ::

Reviewer: Let's remove the instructions to installing zksync-cli and instead link to the Getting Started page for ZKsync CLI.

Reviewer: Instead remove this Install ZKsync CLI section and make it a part of the Prerequisites.

The downloaded project is already configured and contains all the required plugins.
::

**Configure Hardhat:**

- Open your project's `hardhat.config.ts` file and configure it according to your project needs. This includes setting up ZKsync network
   endpoints and compiler versions.

**Set up your environment:**

- Rename the `.env.example` file to `.env` and add your wallet’s private key:

     ```text
     WALLET_PRIVATE_KEY=YourPrivateKeyHere
     ```

::callout{icon="i-heroicons-exclamation-triangle" color="amber"}
For development and testing, it is recommended to use burner wallets. Avoid using real private keys to prevent security risks.
::

**Install project dependencies:**

- Navigate to your project directory and install dependencies:

::code-group

```bash [yarn]
  yarn install
```

```bash [npm]
  npm install
```

::

**Compile your contracts:**

- Before deploying, compile your contracts to ensure they are ready for ZKsync:

::code-group

```bash [yarn]
yarn hardhat compile
```

```bash [npx]
npx hardhat compile
```

::

**Deploy and test:**

> - Use Hardhat scripts to deploy your contracts to ZKsync Era. Refer to the [Getting Started](/build/tooling/hardhat/guides/getting-started)

Reviewer: I am not sure if this page adds much to how to setup a hardhat zksync project. This doesn't describe how to install hardhat-zksync to a project, it explains how to use a project that already has it without explaining what was added to make it a hardhat-zksync project.
We should have this guide have a notice that you can use the existing templates with ZKsync CLI, but the instructions on this page should show how to set up a project without it.

::callout{icon="i-heroicons-light-bulb"}
If you're migrating an existing hardhat project to `zksync-era`, please refer to
the [migration guide](/build/tooling/hardhat/guides/migrating-to-zksync) for
detailed instructions.
::

### Installation steps

**Install ZKsync CLI:**

- The ZKsync CLI simplifies project setup and local testing. Install it globally with npm or yarn:

::code-group

  ```bash [yarn]
    yarn global add zksync-cli
  ```

  ```bash [npm]
    npm install -g zksync-cli
  ```

::

**Create a new project:**

- Use the ZKsync CLI to create a new Hardhat project. Choose the template based on your contract language (Solidity or Vyper):

#### Solidity project

```bash
npx zksync-cli create demo --template hardhat_solidity
```

#### Vyper project

```sh
npx zksync-cli create demo --template hardhat_vyper
```

::callout{icon="i-heroicons-light-bulb"}
Replace `<project-name>` with your desired project name. This command creates a `demo` folder and clones a Hardhat template project inside it.
The downloaded project is already configured and contains all the required plugins.

> ::

Reviewer: Where is this `<project-name>`? Instead of making this a callout, move this description up at the beginning of this section. This is too late to describe what happened with the commands above.

**Configure Hardhat:**

> - Open your project's `hardhat.config.ts` file and configure it according to your project needs. This includes setting up ZKsync network

Reviewer: Isn't this already configured in our template when you use ZKsync CLI? This shouldn't be a step, it should describe like it currently does that the hardhat config has configuration settings for ZKsync.

   endpoints and compiler versions.

**Set up your environment:**

- Rename the `.env.example` file to `.env` and add your wallet’s private key:

     ```text
     WALLET_PRIVATE_KEY=YourPrivateKeyHere
     ```

::callout{icon="i-heroicons-exclamation-triangle" color="amber"}
For development and testing, it is recommended to use burner wallets. Avoid using real private keys to prevent security risks.
::

**Install project dependencies:**

- Navigate to your project directory and install dependencies:

::code-group

```bash [yarn]
  yarn install
```

```bash [npm]
  npm install
```

> ::

Reviewer: ZKsync CLI already runs the installation as part of the creation setup.

**Compile your contracts:**

> - Before deploying, compile your contracts to ensure they are ready for ZKsync:

Reviewer: ```suggestion
- Before deploying, compile your contracts:
```

::code-group

```bash [yarn]
yarn hardhat compile
```

```bash [npx]
npx hardhat compile
```

::

**Deploy and test:**

> - Use Hardhat scripts to deploy your contracts to ZKsync Era.

Reviewer: ```suggestion
- Use Hardhat scripts to deploy your contracts.
```

We don't know what network they are deploying to. By default it is ZKsync Era, but they can change it to a different network like a ZK Chain.
